### ISSUE TITLE
[jvm-packages] Executors fail during prediction (Doing rabit call after Finalize)

### ISSUE THREAD

**Issue by titicaca** [dmlc/xgboost#1284]
I was trying to predict data using the trained model in a spark cluster of 3 nodes with yarn, but executors failed by accident.  New executors will be started to continue the spark task. The spark task can be finished sucessfully and I can still get the correst predictions. However, after several times, the spark context will be shut down because the maximum executor failures reached. 

![image](https://cloud.githubusercontent.com/assets/1780746/16175374/3eb863f6-361d-11e6-9445-d8e9a0423760.png)

I tracked the logged in the corresponding yarn container, and found that "Doing rabit call after Finalize" (see the attached log).  Then I traced the code, which is in utils:Check in rabit/src/engine.cc. It seems executors failed because the util:Check failed and return exit(-1). 

It seems rabit is init and shut down in buildDistributedBoosters in train function. Is there any rabit call in prediction? 
Any idea about this problem? 

[Logs for container_1466187764481_0008_01_000007.pdf](https://github.com/dmlc/xgboost/files/322100/Logs.for.container_1466187764481_0008_01_000007.pdf)

**Comment by CodingCat** [comment]
Would you post the complete stack trace of your executor at the moment of executor lost?

Additionally, what's your data size, hardware configuration and xgboost configuration?

**Comment by titicaca** [comment]
Thanks for the quick reply! 

Attached is the corresponding spark log. There are three errors indicating executor fails. I didn't find any information about why the executor failed.  

The testing dataset is iris, which has 150 points with 1 label and 4 features. The cluster has 3 nodes, each node has 1 core and 512M mem for spark job. 

XGBoost Configuration is as follows:
{"objective":"multi:softmax","num_class":3,"eval_metric":"mlogloss","nthread":1,"eta":0.3,"max_depth":2,"gamma":0,"subsample":1,"colsample_bytree":1,"min_child_weight":1,"silent":0} 
In addition, nworker is 3, nround is 10
 
[spark.log.txt](https://github.com/dmlc/xgboost/files/322292/spark.log.txt)

**Comment by CodingCat** [comment]
I'm guessing that the container is killed by YARN...would you also post NodeManager's log?

**Comment by titicaca** [comment]
Yes, the yarn container is killed. The exist code is 255. I checked the C code  in utils:Check in rabit/src/engine.cc, the check failed the return exit(-1), could it be the reason? 

Attached is the detailed NodeManager's Log.

[yarn-yarn-nodemanager-vm68.txt](https://github.com/dmlc/xgboost/files/324804/yarn-yarn-nodemanager-vm68.txt)
 

2016-06-21 18:25:14,479 WARN org.apache.hadoop.yarn.server.nodemanager.DefaultContainerExecutor: Exception from container-launch with container ID: container_1466502817260_0002_01_000004 and exit code: 255
ExitCodeException exitCode=255: 
	at org.apache.hadoop.util.Shell.runCommand(Shell.java:538)
	at org.apache.hadoop.util.Shell.run(Shell.java:455)
	at org.apache.hadoop.util.Shell$ShellCommandExecutor.execute(Shell.java:702)
	at org.apache.hadoop.yarn.server.nodemanager.DefaultContainerExecutor.launchContainer(DefaultContainerExecutor.java:195)
	at org.apache.hadoop.yarn.server.nodemanager.containermanager.launcher.ContainerLaunch.call(ContainerLaunch.java:300)
	at org.apache.hadoop.yarn.server.nodemanager.containermanager.launcher.ContainerLaunch.call(ContainerLaunch.java:81)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)

**Comment by CodingCat** [comment]
hmmm...from the log, it seems that the error happens only after the training is done, i.e. prediction phase

I haven't got the exact reason of the failure, but what's the version of Spark are you using, it seems that you are using something earlier than 1.6?

**Comment by titicaca** [comment]
I'm using Spark 1.5.1. Could it be the reason? If yes, how can I fix it ?  Is rabit needed in prediction phase??  I noticed that the rabit is initialized and finalized in the training phase.

By the way, in the spark local model, it can work without such error.

**Comment by CodingCat** [comment]
rabit shall not be involved in prediction phase

**Comment by CodingCat** [comment]
could you try spark 1.6.x?

**Comment by titicaca** [comment]
The spark version in my cluster is 1.5.1.  Could you give me some clues to fix this bug? I'm quite confused why  "Doing rabit call after Finalize" happened in the prediction phase.

**Comment by CodingCat** [comment]
I looked at the log again, after all training is done, `16/06/20 04:13:58 INFO org.apache.spark.scheduler.DAGScheduler: Submitting 6 missing tasks from ResultStage 16 (MapPartitionsRDD[15] at map at XGBoost.scala:44)` appeared again, would you check why the training job is submitted again?

**Comment by CodingCat** [comment]
BTW, I was not able to map XGBoost.scala:L44 to any working code....would you pull the latest code of xgboost, try again and upload the log?

**Comment by titicaca** [comment]
XGBoost.scala:44 is predict method,  which is,
`val res = xgb.predict(testRDD, false).flatMap(x => x).map(_.mkString(delimiter))`

Thus it is the prediction job by using the trained model. 

Attached file is XGBoost.scala
[XGBoost.scala.txt](https://github.com/dmlc/xgboost/files/332914/XGBoost.scala.txt)

Thanks

**Comment by titicaca** [comment]
Hi @CodingCat ,

I found that, the executors won't fail, when I set a system sleep of 60 seconds between the training and prediction phase. Do you know why? Maybe something with rabit? 

Attached is the updated spark log, where I tried two times of training + prediction.
In the first time, prediction is after training immediately and the executors failed. In the second time, the prediction waited for 60s seconds after training, and the task finished successfully.


[discover-spark.log.txt](https://github.com/dmlc/xgboost/files/334785/discover-spark.log.txt)

**Comment by CodingCat** [comment]
sorry for the late response, I was in travelling 

Did you change the spark version number in pom.xml when compiling xgboost for your 1.5.1 cluster, if not, I think it might be unpredictable what will happen in the program

**Comment by CodingCat** [comment]
for 60s sleeping issue, I have no idea about what happened since nodemanager only killed "died" executor

**Comment by titicaca** [comment]
Yes, the version is changed to 1.5.1 in pom when compiling .

If model prediction starts immediately after the model training: 
  xgboost “Doing rabit call after Finalize”， and exit -1
  Then executors process exited and are killed by nodemanager

I think it must be somewhere using rabit, but rabit is already finalized in the training phase. 

The following codes are what I found in rabit, the check failed and exited-1 which killed the jvm process:

```
/*! \brief singleton method to get engine */
IEngine *GetEngine() {
  // un-initialized default manager.
  static AllreduceBase default_manager;
  ThreadLocalEntry* e = EngineThreadLocal::Get();
  IEngine* ptr = e->engine.get();
  if (ptr == nullptr) {
    utils::Check(e->initialized,
                 "Doing rabit call after Finalize");
    return &default_manager;
  } else {
    return ptr;
  }
}

inline void Check(bool exp, const char *fmt, ...) {
  if (!exp) {
    std::string msg(kPrintBuffer, '\0');
    va_list args;
    va_start(args, fmt);
    vsnprintf(&msg[0], kPrintBuffer, fmt, args);
    va_end(args);
    HandleCheckError(msg.c_str());
  }
}

inline void HandleCheckError(const char *msg) {
  fprintf(stderr, "%s\n", msg);
  exit(-1);
}
```

Answers for the following questions are appreciate and helpful for me to locate the problem:
1. I want to know why rabit is called in the prediction phase. Is it possible if the trained model is lost, and reload from the cache which call rabit?
2. I noticed that, the booster need to be broadcast by spark broadcast operator before predction. Does it have any conflits with rabit broadcast? Because I found that the broadcast calls rabit engine
```
// broadcast data to all other nodes from root
inline void Broadcast(void *sendrecv_data, size_t size, int root) {
  engine::GetEngine()->Broadcast(sendrecv_data, size, root);
}
```

**Comment by CodingCat** [comment]
>1. I want to know why rabit is called in the prediction phase. Is it possible if the trained model is lost, and reload from the cache which call rabit?

As the expected case, rabit is not called in the prediction phase, 

test and train phase shall not have direct interaction, since the booster is collected as a driver-side variable before broadcasting 

>>> 2. I noticed that, the booster need to be broadcast by spark broadcast operator before predction. Does it have any conflits with rabit broadcast? I found that the broadcast calls rabit engine, which may cause the problem if the rabit broadcast happened in prediction phase.

no, broadcast of booster is an Spark-domain functionality

**Comment by titicaca** [comment]
Okay.. But here the weird is that the log shows rabit is called after finalized..

**Comment by CodingCat** [comment]
have you ever run `mvn test` successfully? (with/without changing spark version to 1.5.1)

**Comment by titicaca** [comment]
Yes, both spark 1.6.1 and 1.5.1 can pass mvn test

**Comment by titicaca** [comment]
May I ask why the condition `if(testSamples.hasNext)` in spark predict is like this? Shall it be `if(testSamples.isEmpty)` ?

```
testSet.mapPartitions { testSamples =>
      if (testSamples.hasNext) {
        val cacheFileName = {
          if (useExternalCache) {
            s"$appName-dtest_cache-${TaskContext.getPartitionId()}"
          } else {
            null
          }
        }
        val dMatrix = new DMatrix(new JDMatrix(testSamples, cacheFileName))
        Iterator(broadcastBooster.value.predict(dMatrix))
      } else {
        Iterator()
      }
    }
```

**Comment by CodingCat** [comment]
![image](https://cloud.githubusercontent.com/assets/678008/16585606/b1d79b94-428f-11e6-81a0-5eed72f31e5b.png)
They are the same in implementation

**Comment by titicaca** [comment]
I see. Thanks. 

May I ask one more question about rabit. Do we have to repartition the trainRDD to numWorkers partitions?  As I understood, each partition corresponds to a task, which is executed in one thread. Can each thread has its own rabit? Or rabit has to be used (initialized and finalized) in process level for each executor?

**Comment by CodingCat** [comment]
each thread (i.e. task or xgboost worker) can have its own rabit.....

**Comment by titicaca** [comment]
I see, THX!

By the way, I think I have found the bug. The prediction do call rabit engine in distributed model. In c_api.cc, XGBoosterPredict calls bst->LazyInit(), 
```
inline void LazyInit() {
    if (!configured_) {
      learner_->Configure(cfg_);
      configured_ = true;
    }
    if (!initialized_) {
      learner_->InitModel();
      initialized_ = true;
    }
  }
```
configured_ in the Booster Constructor is false and will call `learner_->Configure(cfg_);`
In Configure() in learner.cc, `rabit::IsDistributed()` calls rabit::getEngine() and cause the error: "Doing rabit call after Finalize"
```
if (tparam.dsplit == 0 && rabit::IsDistributed()) {
      tparam.dsplit = 2;
    }
```

I just created a new pull request for this bug, where I added rabit init also in spark prediction phase. It is tested in yarn-cluster and executors won't fail in prediction phase anymore.
